feat: Update InvoicingEngine to receive rules instead of file_name

adrianmoloce opened the pull request:
- Engine now takes an optional rules
  - If rules are not provided use default basic.json rules
  - Update tests to reflect changes
  - Add todo because spoted logic error in _process_item, vat not used

amoloce committed:
feat: Update InvoicingEngine to receive rules instead of file_name

  - Engine now takes an optional rules
  - If rules are not provided use default basic.json rules
  - Update tests to reflect changes
  - Add todo because spoted logic error in _process_item, vat not used

koosie0507 commented on tests/invoice_utils/engine/rule_engine_basic_input_processing_test.py:
```diff
@@ -11,7 +11,7 @@ def test_process_empty_input_data_outputs_base_invoice_document_structure(
     expected_invoice_date = datetime(2022, 1, 15, 13, 14, 15)
 
-    output = InvoicingEngine(resolve_path("empty.json")).process(
+    output = InvoicingEngine([{}]).process(
```
*nit: `[]` would be the quivalent of `empty.json`.

koosie0507 commented on tests/invoice_utils/conftest.py:
```diff
@@ -33,6 +33,11 @@ def f(name: str) -> str:
 
 
+@pytest.fixture
+def test_basic_rules(read_text):
```
`test_*` is the convention for test names, not fixture names. please rename to `basic_rules`.

koosie0507 commented on src/invoice_utils/engine/_engine.py:
```diff
@@ -33,6 +25,17 @@ def __init_invoice(self):
         }
 
+    @staticmethod
+    def __set_default_rules():
```
This class accepts the `rules` directly. This `@staticmethod` doesn't belong here anymore because i/o results in the `rules` being passed to instances of this class. 
One other concern is that this method  causes the `InvoicingEngine` initializer to fail. The calling code fails implicitly. Since explicit is always better than implicit, failing explicitly while determining the rule template file to use is preferred.

koosie0507 commented on src/invoice_utils/engine/_engine.py:
```diff
@@ -4,27 +4,19 @@
+
 class InvoicingEngine:
-    def __init__(self, file_name: str):
+    def __init__(self, rules: Optional[list[dict]] = None):
```
Please make this parameter mandatory (i.e. `list[dict]`) and move setting the default from the initialiser to the calling code.

koosie0507 commented on src/invoice_utils/engine/_engine.py:
```diff
@@ -33,6 +25,17 @@ def __init_invoice(self):
+    def __set_default_rules():
+        fpath = pathlib.Path(__file__).parent.parent / "basic.json"
+        if not fpath.exists():
+            raise InvoicingInputError("basic.json")
```
In the case of `basic.json`, the web app shouldn't start up if this file is missing or invalid. That check belongs in the entry point of the app or in the app settings.

amoloce committed:
rename test_basic_rules fixture to basic_rules

amoloce committed:
make rules a mandatory param in InvoicingEngine

amoloce committed:
move set default rules logic from engine to calling code

amoloce committed:
sync data from empty.json with engine input

amoloce committed:
add pydantic-setting dependency

amoloce committed:
check default rules file exist or raise error

amoloce committed:
remove unused imports in engine

adrianmoloce commented on tests/invoice_utils/test_web.py:
```diff
@@ -326,3 +325,20 @@ def test_template_related_variables_with_bad_inputs_raise_errors(
+        import invoice_utils.web as web
+
+        with pytest.raises(HTTPException):
+            with TestClient(web.app) as client:
```
Note: We can only test the startup event using a context manager [docs](https://fastapi.tiangolo.com/advanced/testing-events/), if I did this in the http fixture, all tests would be impacted by the checks at startup.

koosie0507 replied:
when the rule template specified by `INVOICE_UTILS_RULE_TEMPLATE_NAME` environment variable name does not exist, We should fall back to the `DEFAULT_RULE_TEMPLATE_NAME` constant.
We can achieve this coercion by using the `template_repository()` dependency in a [lifespan function](https://fastapi.tiangolo.com/advanced/events/#lifespan) for the FastAPI app.
This assumes that the actual rule name being used is mutable and changed during app startup accordingly.

adrianmoloce replied:
@koosie0507 the behaviour is the same for the lifespan route, the tests that do not use TestClient in a context manager (so all tests using http) do not actually make use of the lifespan code.
I tested that separately in 2 tests which specifically test that behaviour.

amoloce committed:
make startup event test more explicit for rule

adrianmoloce commented on src/invoice_utils/engine/_engine.py:
```diff
@@ -57,6 +44,7 @@ def _process_item(self, item_no: int, item_tax: Decimal, item: InvoicedItem):
                     "name": name,
                     "value": round(val + item_price, 6)
                 })
+        # TODO: This line overwrites the item_tax mandatory parameter received
```
@koosie0507 Can you check this out? at line 122, we create a vat variable given as the item_tax parameter, that is not used since here at line 48 we use a different logic.

koosie0507 replied:
This line allows overwriting the default VAT. Don't change it please. Also, please see my other comment about #TODO comments.

koosie0507 commented on src/invoice_utils/web.py:
```diff
@@ -105,6 +115,13 @@ def generate_invoice(request: InvoiceRequest):
 
+def _get_default_rules(root_dir):
+    basic_rules = root_dir / "basic.json"
+    # TODO: Change this in phase 2 by using repository
```
Please remove this comment. Add an issue in the tracker or to a personal reminders software instead of this comment here. Inline comments should be useful information about the code that _already exists_ in the repository. They usually explain complex algorithms and/or design choices.

koosie0507 commented on tests/invoice_utils/test_web.py:
```diff
@@ -326,3 +325,24 @@ def test_template_related_variables_with_bad_inputs_raise_errors(
+    ],
+    indirect=["environment"]
+)
+def test_app_does_not_start_with_bad_default_rule_template(
```
Is there a test that checks that the app doesn't start if the `DEFAULT_RULE_TEMPLATE_NAME` doesn't exist? That is to say, if `template_repo.exists(DEFAULT_RULE_TEMPLATE_NAME)` returns `False` then the `app` shouldn't start at all.

adrianmoloce replied:
Shouldn't the `app` fail only if the `template_repo.exists(DEFAULT_RULE_TEMPLATE_NAME)` AND then the default "basic.json" file doesn't exist?

koosie0507 commented on src/invoice_utils/config.py:
```diff
@@ -28,3 +29,8 @@ def _str_to_bool(arg: str) -> bool:
     "INVOICE_UTILS_TEMPLATES_DIR",
     DEFAULT_BODY_TEMPLATE_DIRECTORY
 )
+INVOICE_UTILS_DEFAULT_RULE_TEMPLATE_NAME = os.getenv(
```
This variable should be called `INVOICE_UTILS_RULE_TEMPLATE_NAME`. The `DEFAULT_...` is specified as a constant above.

amoloce committed:
remove line comments and rename rule template name

amoloce committed:
fix: Solve PR comments

  - Change onevent to lifespan for the mandatory app config
  - Unit tests for lifespan
  - Update /invoice endpoint to use repository to get the rules
  - Unit test the endpoint and it flows
  - Rename config variables more appropriately.
  - Change basic.json structure

amoloce committed:
extract single use parameter

amoloce committed:
Add a few more test cases
